File: src/adapters/golang/access/access.go
package access

import (
	"context"
	"crypto/tls"
	"errors"
	"fmt"
	"os"
	"strings"
	"time"

	"authguard/adapters/golang/model"
	"authguard/adapters/golang/util"
	"google.golang.org/grpc"
	"google.golang.org/grpc/credentials"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/protobuf/types/known/wrapperspb"
)

const resolveScopeMethod = "/authguard.access.v1.AccessContextService/ResolveScope"

const (
	GRPCTargetEnv           = "AUTHGUARD_GRPC_TARGET"
	GRPCTLSEnv              = "AUTHGUARD_GRPC_TLS"
	AccessContextHMACKeyEnv = "AUTHGUARD_ACCESS_CONTEXT_HMAC_KEY"
)

var (
	ErrAccessContextUnavailable = errors.New("authguard access context is not available")
	ErrScopeResolverUnavailable = errors.New("no resolver is configured for the Authguard scope token")
)

type ActionMismatchError struct {
	Expected string
	Actual   string
}

func (e ActionMismatchError) Error() string {
	return "authguard action mismatch: expected `" + e.Expected + "`, got `" + e.Actual + "`"
}

type HeaderReader interface {
	Header(name string) string
}

type IAccessContextResolver interface {
	Resolve(ctx context.Context, headers HeaderReader) (*model.RequestAccess, error)
}

type ScopeTokenClient interface {
	ResolveScope(ctx context.Context, token string) (string, error)
}

type GRPCScopeTokenClient struct {
	connection *grpc.ClientConn
}

func NewGRPCScopeTokenClient(target string, options ...grpc.DialOption) (*GRPCScopeTokenClient, error) {
	connection, err := grpc.NewClient(target, options...)
	if err != nil {
		return nil, fmt.Errorf("create Authguard scope resolver client: %w", err)
	}
	return &GRPCScopeTokenClient{connection: connection}, nil
}

func NewGRPCScopeTokenClientFromEnv() (*GRPCScopeTokenClient, error) {
	target := strings.TrimSpace(os.Getenv(GRPCTargetEnv))
	if target == "" {
		return nil, fmt.Errorf("%s is required", GRPCTargetEnv)
	}
	tlsEnabled, err := parseBool(defaultString(os.Getenv(GRPCTLSEnv), "false"))
	if err != nil {
		return nil, fmt.Errorf("parse %s: %w", GRPCTLSEnv, err)
	}
	transportCredentials := credentials.TransportCredentials(insecure.NewCredentials())
	if tlsEnabled {
		transportCredentials = credentials.NewTLS(&tls.Config{MinVersion: tls.VersionTLS12})
	}
	util.LogDebug(context.Background(), "authguard.scope_token.grpc.configured",
		"resolver_mode", "grpc", "tls", tlsEnabled)
	return NewGRPCScopeTokenClient(
		target,
		grpc.WithTransportCredentials(transportCredentials),
	)
}

func (c *GRPCScopeTokenClient) ResolveScope(ctx context.Context, token string) (string, error) {
	started := time.Now()
	util.LogDebug(ctx, "authguard.scope_token.grpc.started", "resolver_mode", "grpc")
	response := new(wrapperspb.StringValue)
	if err := c.connection.Invoke(ctx, resolveScopeMethod, wrapperspb.String(token), response); err != nil {
		util.LogWarn(ctx, "authguard.scope_token.grpc.failed",
			"resolver_mode", "grpc", "error_category", util.ErrorCategory(err),
			"duration_ms", time.Since(started).Milliseconds())
		return "", fmt.Errorf("resolve Authguard scope token: %w", err)
	}
	util.LogDebug(ctx, "authguard.scope_token.grpc.succeeded",
		"resolver_mode", "grpc", "duration_ms", time.Since(started).Milliseconds())
	return response.Value, nil
}

func (c *GRPCScopeTokenClient) Close() error {
	return c.connection.Close()
}

// HeaderAccessContextResolver verifies and decodes the context injected by Envoy.
type HeaderAccessContextResolver struct {
	signingKey string
}

func NewHeaderAccessContextResolver(signingKey string) (HeaderAccessContextResolver, error) {
	if _, err := util.SignEncodedAccessContext("probe", signingKey); err != nil {
		return HeaderAccessContextResolver{}, err
	}
	return HeaderAccessContextResolver{signingKey: signingKey}, nil
}

func NewHeaderAccessContextResolverFromEnv() (HeaderAccessContextResolver, error) {
	signingKey := os.Getenv(AccessContextHMACKeyEnv)
	if signingKey == "" {
		return HeaderAccessContextResolver{}, fmt.Errorf("%s is required", AccessContextHMACKeyEnv)
	}
	return NewHeaderAccessContextResolver(signingKey)
}

func (r HeaderAccessContextResolver) Resolve(
	_ context.Context,
	headers HeaderReader,
) (*model.RequestAccess, error) {
	encoded := headers.Header(util.AccessContextHeader)
	if encoded == "" {
		return nil, nil
	}
	signingKey := r.signingKey
	if signingKey == "" {
		signingKey = os.Getenv(AccessContextHMACKeyEnv)
	}
	if signingKey == "" {
		return nil, fmt.Errorf("%s is required", AccessContextHMACKeyEnv)
	}
	accessContext, err := util.VerifySignedAccessContext(encoded, signingKey)
	if err != nil {
		return nil, err
	}
	requestAccess := accessContext.RequestAccess()
	return &requestAccess, nil
}

// GRPCAccessContextResolver exchanges a request-scoped opaque token for the
// complete access context through Authguard's gRPC API.
type GRPCAccessContextResolver struct {
	client ScopeTokenClient
}

func NewGRPCAccessContextResolver(client ScopeTokenClient) GRPCAccessContextResolver {
	return GRPCAccessContextResolver{client: client}
}

func NewGRPCAccessContextResolverFromEnv() (GRPCAccessContextResolver, error) {
	client, err := NewGRPCScopeTokenClientFromEnv()
	if err != nil {
		return GRPCAccessContextResolver{}, err
	}
	return NewGRPCAccessContextResolver(client), nil
}

func (r GRPCAccessContextResolver) Close() error {
	if closer, ok := r.client.(interface{ Close() error }); ok {
		return closer.Close()
	}
	return nil
}

func (r GRPCAccessContextResolver) Resolve(
	ctx context.Context,
	headers HeaderReader,
) (*model.RequestAccess, error) {
	token := headers.Header(util.ScopeTokenHeader)
	if token == "" {
		return nil, nil
	}
	if r.client == nil {
		return nil, ErrScopeResolverUnavailable
	}
	encoded, err := r.client.ResolveScope(ctx, token)
	if err != nil {
		return nil, err
	}
	accessContext, err := util.DecodeAccessContext(encoded)
	if err != nil {
		return nil, err
	}
	requestAccess := accessContext.RequestAccess()
	return &requestAccess, nil
}

type contextKey struct{}

type contextValue struct {
	access  model.RequestAccess
	present bool
}

func WithGrantSet(ctx context.Context, grants model.AccessGrantSet) context.Context {
	return WithRequestAccess(ctx, model.RequestAccess{Grants: grants})
}

func WithRequestAccess(ctx context.Context, requestAccess model.RequestAccess) context.Context {
	util.LogDebug(ctx, "authguard.access_context.bound",
		"principal_id", util.SafeLogString(requestAccess.PrincipalID),
		"action", util.SafeLogString(requestAccess.Action),
		"allow_count", len(requestAccess.Grants.AllowResourceURNs),
		"deny_count", len(requestAccess.Grants.DenyResourceURNs))
	return context.WithValue(ctx, contextKey{}, contextValue{access: requestAccess, present: true})
}

func WithoutGrantSet(ctx context.Context) context.Context {
	if current, ok := RequestAccess(ctx); ok {
		util.LogDebug(ctx, "authguard.access_context.cleared",
			"principal_id", util.SafeLogString(current.PrincipalID),
			"action", util.SafeLogString(current.Action))
	}
	return context.WithValue(ctx, contextKey{}, contextValue{})
}

func GrantSet(ctx context.Context) (model.AccessGrantSet, bool) {
	value, ok := ctx.Value(contextKey{}).(contextValue)
	if !ok || !value.present {
		return model.AccessGrantSet{}, false
	}
	return value.access.Grants, true
}

func RequestAccess(ctx context.Context) (model.RequestAccess, bool) {
	value, ok := ctx.Value(contextKey{}).(contextValue)
	if !ok || !value.present {
		return model.RequestAccess{}, false
	}
	return value.access, true
}

func RequireGrantSet(ctx context.Context) (model.AccessGrantSet, error) {
	grants, ok := GrantSet(ctx)
	if !ok {
		util.LogDebug(ctx, "authguard.access_context.required_missing")
		return model.AccessGrantSet{}, ErrAccessContextUnavailable
	}
	return grants, nil
}

func RequireRequestAccess(ctx context.Context) (model.RequestAccess, error) {
	requestAccess, ok := RequestAccess(ctx)
	if !ok {
		util.LogDebug(ctx, "authguard.access_context.required_missing")
		return model.RequestAccess{}, ErrAccessContextUnavailable
	}
	return requestAccess, nil
}

func CurrentScope(ctx context.Context, mapping model.ResourceSQLMapping) (model.SqlScope, error) {
	grants, err := RequireGrantSet(ctx)
	if err != nil {
		return model.SqlScope{}, err
	}
	return util.CompileScope(mapping, grants.AllowResourceURNs, grants.DenyResourceURNs)
}

func CurrentScopeForAction(
	ctx context.Context,
	expectedAction string,
	mapping model.ResourceSQLMapping,
) (model.SqlScope, error) {
	requestAccess, err := RequireRequestAccess(ctx)
	if err != nil {
		return model.SqlScope{}, err
	}
	return ScopeForAction(requestAccess, expectedAction, mapping)
}

func ScopeForAction(
	requestAccess model.RequestAccess,
	expectedAction string,
	mapping model.ResourceSQLMapping,
) (model.SqlScope, error) {
	if requestAccess.Action != expectedAction {
		util.LogDebug(context.Background(), "authguard.sql_scope.action_mismatch",
			"principal_id", util.SafeLogString(requestAccess.PrincipalID),
			"expected_action", util.SafeLogString(expectedAction),
			"actual_action", util.SafeLogString(requestAccess.Action))
		return model.SqlScope{}, ActionMismatchError{
			Expected: expectedAction,
			Actual:   requestAccess.Action,
		}
	}
	return util.CompileScope(
		mapping,
		requestAccess.Grants.AllowResourceURNs,
		requestAccess.Grants.DenyResourceURNs,
	)
}

func parseBool(value string) (bool, error) {
	switch strings.ToLower(strings.TrimSpace(value)) {
	case "1", "true", "yes", "on":
		return true, nil
	case "0", "false", "no", "off", "":
		return false, nil
	default:
		return false, fmt.Errorf("expected boolean, got %q", value)
	}
}

func defaultString(value string, fallback string) string {
	value = strings.TrimSpace(value)
	if value == "" {
		return fallback
	}
	return value
}

File: src/adapters/python/authguard_adapter/access.py
from __future__ import annotations

from contextvars import ContextVar, Token
import os
import time
from typing import Protocol

import grpc
from google.protobuf.wrappers_pb2 import StringValue

from authguard_adapter.model import AccessContext, AccessGrantSet, RequestAccess

GRPC_TARGET_ENV = "AUTHGUARD_GRPC_TARGET"
GRPC_TLS_ENV = "AUTHGUARD_GRPC_TLS"
ACCESS_CONTEXT_HMAC_KEY_ENV = "AUTHGUARD_ACCESS_CONTEXT_HMAC_KEY"


class AccessContextUnavailable(RuntimeError):
    pass


class AccessHeaders(Protocol):
    def header(self, name: str) -> str | None:
        ...


class IAccessContextResolver(Protocol):
    def resolve(self, headers: AccessHeaders) -> RequestAccess | None:
        ...


class ScopeTokenClient(Protocol):
    def resolve_scope(self, token: str) -> str:
        ...


class GrpcScopeTokenClient:
    _METHOD = "/authguard.access.v1.AccessContextService/ResolveScope"

    def __init__(self, target: str, *, secure: bool = False) -> None:
        self._channel = (
            grpc.secure_channel(target, grpc.ssl_channel_credentials())
            if secure
            else grpc.insecure_channel(target)
        )
        self._resolve = self._channel.unary_unary(
            self._METHOD,
            request_serializer=StringValue.SerializeToString,
            response_deserializer=StringValue.FromString,
        )
        from authguard_adapter.util import _log_debug

        _log_debug(
            "authguard.scope_token.grpc.configured",
            resolver_mode="grpc",
            tls=secure,
        )

    def resolve_scope(self, token: str) -> str:
        from authguard_adapter.util import _log_debug, _log_warning, error_category

        started = time.perf_counter()
        _log_debug("authguard.scope_token.grpc.started", resolver_mode="grpc")
        try:
            response: StringValue = self._resolve(StringValue(value=token))
        except grpc.RpcError as error:
            _log_warning(
                "authguard.scope_token.grpc.failed",
                resolver_mode="grpc",
                error_category=error_category(error),
                duration_ms=int((time.perf_counter() - started) * 1000),
            )
            raise
        _log_debug(
            "authguard.scope_token.grpc.succeeded",
            resolver_mode="grpc",
            duration_ms=int((time.perf_counter() - started) * 1000),
        )
        return response.value

    def close(self) -> None:
        self._channel.close()

    @classmethod
    def from_env(cls) -> GrpcScopeTokenClient:
        target = os.getenv(GRPC_TARGET_ENV, "").strip()
        if not target:
            raise ValueError(f"{GRPC_TARGET_ENV} is required")
        secure = _parse_bool(os.getenv(GRPC_TLS_ENV, "false"), GRPC_TLS_ENV)
        return cls(target, secure=secure)


class HeaderAccessContextResolver:
    """Verifies and decodes the trusted access context injected by Envoy."""

    def __init__(self, signing_key: str | None = None) -> None:
        if signing_key is not None:
            from authguard_adapter.util import sign_encoded_access_context

            sign_encoded_access_context("probe", signing_key)
        self._signing_key = signing_key

    @classmethod
    def from_env(cls) -> HeaderAccessContextResolver:
        signing_key = os.getenv(ACCESS_CONTEXT_HMAC_KEY_ENV, "")
        if not signing_key:
            raise ValueError(f"{ACCESS_CONTEXT_HMAC_KEY_ENV} is required")
        return cls(signing_key)

    def resolve(self, headers: AccessHeaders) -> RequestAccess | None:
        from authguard_adapter.util import ACCESS_CONTEXT_HEADER, verify_signed_access_context

        encoded = headers.header(ACCESS_CONTEXT_HEADER)
        if not encoded:
            return None
        signing_key = self._signing_key or os.getenv(ACCESS_CONTEXT_HMAC_KEY_ENV, "")
        if not signing_key:
            raise ValueError(f"{ACCESS_CONTEXT_HMAC_KEY_ENV} is required")
        context = verify_signed_access_context(encoded, signing_key)
        return context.request_access()


class GrpcAccessContextResolver:
    """Resolves an opaque request scope token through Authguard gRPC."""

    def __init__(self, client: ScopeTokenClient) -> None:
        self._client = client

    def resolve(self, headers: AccessHeaders) -> RequestAccess | None:
        from authguard_adapter.util import SCOPE_TOKEN_HEADER, decode_access_context

        token = headers.header(SCOPE_TOKEN_HEADER)
        if not token:
            return None
        context = decode_access_context(self._client.resolve_scope(token))
        return context.request_access()

    @classmethod
    def from_env(cls) -> GrpcAccessContextResolver:
        return cls(GrpcScopeTokenClient.from_env())

    def close(self) -> None:
        close = getattr(self._client, "close", None)
        if callable(close):
            close()


_current_access: ContextVar[RequestAccess | None] = ContextVar("authguard_current_access", default=None)


def set_current(grants: AccessGrantSet) -> Token[RequestAccess | None]:
    return set_current_access(RequestAccess.from_grants(grants))


def set_current_access(request_access: RequestAccess) -> Token[RequestAccess | None]:
    from authguard_adapter.util import _log_debug

    _log_debug(
        "authguard.access_context.bound",
        principal_id=request_access.principal_id,
        action=request_access.action,
        allow_count=len(request_access.grants.allow_resource_urns),
        deny_count=len(request_access.grants.deny_resource_urns),
    )
    return _current_access.set(request_access)


def reset_current(token: Token[RequestAccess | None]) -> None:
    from authguard_adapter.util import _log_debug

    current = _current_access.get()
    _current_access.reset(token)
    if current is not None:
        _log_debug(
            "authguard.access_context.cleared",
            principal_id=current.principal_id,
            action=current.action,
        )


def clear_current() -> None:
    from authguard_adapter.util import _log_debug

    current = _current_access.get()
    _current_access.set(None)
    if current is not None:
        _log_debug(
            "authguard.access_context.cleared",
            principal_id=current.principal_id,
            action=current.action,
        )


def get_current() -> AccessGrantSet | None:
    request_access = get_current_access()
    return request_access.grants if request_access is not None else None


def get_current_access() -> RequestAccess | None:
    return _current_access.get()


def require_current() -> AccessGrantSet:
    grants = get_current()
    if grants is None:
        from authguard_adapter.util import _log_debug

        _log_debug("authguard.access_context.required_missing")
        raise AccessContextUnavailable("authguard access context is not available")
    return grants


def require_current_access() -> RequestAccess:
    request_access = get_current_access()
    if request_access is None:
        from authguard_adapter.util import _log_debug

        _log_debug("authguard.access_context.required_missing")
        raise AccessContextUnavailable("authguard access context is not available")
    return request_access


def _parse_bool(value: str, name: str) -> bool:
    normalized = value.strip().lower()
    if normalized in {"1", "true", "yes", "on"}:
        return True
    if normalized in {"0", "false", "no", "off"}:
        return False
    raise ValueError(f"{name} must be a boolean")
